playwright/src/main/java/com/microsoft/playwright/impl/UrlMatcher.java

```diff
@@ -68,50 +64,79 @@ static String resolveUrl(URL baseUrl, String spec) {
     }
   }
 
-  UrlMatcher(URL base, String url) {
-    this(url, toPredicate(Pattern.compile(globToRegex(resolveUrl(base, url)))).or(s -> url == null || url.equals(s)));
+  UrlMatcher(URL baseURL, String glob) {
+    this(baseURL, glob, null, null);
   }
 
   UrlMatcher(Pattern pattern) {
-    this(pattern, toPredicate(pattern));
+    this(null, null, pattern, null);
   }
   UrlMatcher(Predicate<String> predicate) {
-    this(predicate, predicate);
+    this(null, null, null, predicate);
   }
 
-  private UrlMatcher(Object rawSource, Predicate<String> predicate) {
-    this.rawSource = rawSource;
+  private UrlMatcher(URL baseURL, String glob, Pattern pattern, Predicate<String> predicate) {
+    this.baseURL = baseURL;
+    this.glob = glob;
+    this.pattern = pattern;
     this.predicate = predicate;
   }
 
   boolean test(String value) {
-    return predicate == null || predicate.test(value);
+    return testImpl(baseURL, pattern, predicate, glob, value);
+  }
+
+  private static boolean testImpl(URL baseURL, Pattern pattern, Predicate<String> predicate, String glob, String value) {
```

Comment:
why another method? We'll later on (in future patches) mutate e.g. the `glob` env var - making it `private static` helps us there not to rewrite the UrlMatcher's instance property itself.